repose-aggregator/services/atom-feed/impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/AtomEntryStreamBuilder.scala

```diff
@@ -0,0 +1,105 @@
+    * making a request to the Atom feed, assign the result of this function to a lazy variable like so:
+    * lazy val myFeed = build("http://my.feed.url", MyAuthenticatedRequestFactory)
+    *
+    * @param baseFeedUrl a static locator pointing to the "head" of an Atom feed (e.g., the subscription document)
+    * @param authenticator a URL processor which authenticates connections
+    * @param authenticationTimeLimit a maximum [[Duration]] to wait on authentication before considering authentication
+    *                                a failure
+    * @return a Scala [[scala.collection.immutable.Stream]] which will yield Atom entries in an Atom feed located at the
+    *         provided baseFeedUrl. The entries will be yielded in the order in which they are read from the feed XML,
+    *         from top to bottom.
+    * @throws AuthenticationException when the authenticator fails to authenticate a request
+    */
+  def build(baseFeedUrl: URL,
+            reposeVersion: String,
+            authenticator: Option[AuthenticatedRequestFactory] = None,
+            authenticationTimeLimit: Duration = 1 second): Stream[Entry] = {
+    val requestId = java.util.UUID.randomUUID().toString
+    MDC.put(TracingKey.TRACING_KEY, requestId)
+    buildR(baseFeedUrl, reposeVersion, authenticator, authenticationTimeLimit, requestId)
+  }
+
+  private def buildR(baseFeedUrl: URL,
+                     reposeVersion: String,
+                     authenticator: Option[AuthenticatedRequestFactory] = None,
+                     authenticationTimeLimit: Duration = 1 second,
+                     requestId: String): Stream[Entry] = {
+    val baseFeedConnection = baseFeedUrl.openConnection()
+
+    val authenticatedConnection = authenticator match {
+      case Some(arf) =>
+        val connectionFuture = Future(arf.authenticateRequest(baseFeedConnection, requestId, reposeVersion))
+        Option(Await.result(connectionFuture, authenticationTimeLimit))
+      case None =>
+        Some(baseFeedConnection)
+    }
+
+    authenticatedConnection match {
+      case Some(urlConnection) =>
+        val tracingHeader = TracingHeaderHelper.createTracingHeader(requestId, "1.1 Repose (Repose/" + reposeVersion + ")", None)
+        urlConnection.setRequestProperty(CommonHttpHeader.TRACE_GUID.toString, tracingHeader)
+
+        val feedInputStream = urlConnection.getInputStream
+        val feed = parser.parse[Feed](feedInputStream).getRoot
+        feedInputStream.close()
+
+        feed.getLinks.find(link => link.getRel.equals("next")) match {
+          case Some(nextPageLink) =>
+            feed.getEntries.toStream #::: buildR(nextPageLink.getResolvedHref.toURL, reposeVersion, authenticator, authenticationTimeLimit, requestId)
+          case None =>
+            feed.getEntries.toStream
+        }
+      case None =>
+        throw AuthenticationException
```

Comment:
Oh, right, we could absolutely do a single retry. I think that's a different issue, though? Take the following example:

The FeedReader is kicked by the Akka scheduler. It reads the first page of a feed no problem, so all of those entries are send to associated listeners. Now it goes to read the second page, but the network is terribly noisy, so the request times out. Exception. Well the highWaterMark hasn't been updated because an Exception occurred, but we *did* send out some new entries. Of course, we can't set the high water mark to any of the entries we sent, because then we potentially be skipping older entries. If we don't set the high water mark, though, we are at risk of sending duplicate entries the next time the FeedReader attempts to read the feed. So what do we do?

If we were reading from the bottom of the feed to the top, we could simply set the high water mark after each entry we read. That doesn't work if we read from top to bottom (the natural order), though. The problem is, we don't know if a page will fail until we've read that page. The only solution I can think of, then, is to read all of the pages, buffer them, then send out the entries on a success. That slows the read top-to-bottom case down to the speed of the read bottom-to-top case, though. Both cases become read everything, then send entries (in whatever order).